python/lsst/ip/diffim/subtractImages.py

```diff
@@ -617,6 +605,232 @@ def _sourceSelector(self, sources):
+        sources : `lsst.afw.table.SourceCatalog`
+            Identified sources on the science exposure. This catalog is used to
+            select sources in order to perform the AL PSF matching on stamp
+            images around them.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+            ``scoreExposure`` : `lsst.afw.image.ExposureF`
+                The image difference.
+            ``matchedTemplate`` : `lsst.afw.image.ExposureF`
+                The warped and PSF-matched template exposure.
+        """
+
+        selectSources = self.prepareInputs(template, science, sources,
+                                           finalizedPsfApCorrCatalog=finalizedPsfApCorrCatalog)
+
+        subtractResults = self.runConvolveTemplate(template, science, selectSources)
+
+        return subtractResults
+
+    def runConvolveTemplate(self, template, science, selectSources):
+        """Convolve the template image with a PSF-matching kernel and subtract
+        from the science image.
+
+        Parameters
+        ----------
+        template : `lsst.afw.image.ExposureF`
+            Template exposure, warped to match the science exposure.
+        science : `lsst.afw.image.ExposureF`
+            Science exposure to subtract from the template.
+        selectSources : `lsst.afw.table.SourceCatalog`
+            Identified sources on the science exposure. This catalog is used to
+            select sources in order to perform the AL PSF matching on stamp
+            images around them.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+
+            ``difference`` : `lsst.afw.image.ExposureF`
+                Result of subtracting template and science.
+                Attached PSF is that of the science image.
+            ``matchedTemplate`` : `lsst.afw.image.ExposureF`
+                Warped and PSF-matched template exposure.
+                Attached PSF is that of the science image.
+            ``backgroundModel`` : `lsst.afw.math.Function2D`
+                Background model that was fit while solving for the PSF-matching kernel
+            ``psfMatchingKernel`` : `lsst.afw.math.Kernel`
+                Kernel used to PSF-match the template to the science image.
+        """
+
+        bbox = lsst.geom.Box2I(science.getBBox().getBegin(), science.getBBox().getDimensions())
+        bbox.grow(-20)
+        kernelSources = self.makeKernel.selectKernelSources(template[bbox], science[bbox],
+                                                            candidateList=selectSources,
+                                                            preconvolved=True)
+
+        # creating a kernel to run preconvolution with its own psf
+        # TODO: we need to mirror the kernel in order to get correct cross correlation
+        scienceKernel = science.psf.getKernel()
+
+        matchedScience = self._convolveExposure(science, scienceKernel, self.convolutionControl)
```

Comment:
I'd remove the empty line above this, since these two lines go together logically.